# 2CS SIQ2-SIL2 TP04. Naïve Bayes

Dans ce TP, nous allons traiter Naïve Bayes. C'est le seul algorithme dans notre programme qui crée un modèle génératif (en plus des auto-encodeurs).

- **Binôme 01** : TOUHAR Afnane
- **Binôme 02** : GOUASMIA Malak 
- **Groupe** : SIL2

In [5]:
import matplotlib
import numpy             as np
import pandas            as pd 
import matplotlib.pyplot as plt 
%matplotlib inline

np        .__version__ , \
pd        .__version__ , \
matplotlib.__version__

('1.22.4', '1.5.3', '3.7.1')

In [7]:
from typing import Tuple, List, Dict

**RAPPEL**

Tout le monde connait le théorème de Bayes pour calculer la probabilité conditionnelle d'un évennement $A$ sachant un autre $B$ (si vous ne le connaissez pas; vous n'appartenez pas à tout le monde) : 
$$ P(A|B) = \frac{P(A)P(B|A)}{P(B)}$$

Pour appliquer ce théorème sur un problème d'appentissage automatique, l'idée est simple ; Etant donné une caractéristique $f$ et la sortie $y$ qui peut avoir la classe $c$ : 
- Remplacer $A$ par $y=c$
- Remplacer $B$ par $f$ 
On aura l'équation : 
$$ P(y=c|f) = \frac{P(y=c)P(f|y=c)}{P(f)}$$

On appelle : 
- $P(y=c|f)$ postérieure 
- $P(y=c)$ antérieure
- $P(f|y=c)$ vraisemblance
- $P(f)$ évidence 

Ici, on estime la probablité d'une classe $c$ sachant une caractéristique $f$ en utilisant des données d'entrainement. Maintenant, on veut estimer la probabilité d'une classe $c$ sachant un vecteur de caractéristiques $\overrightarrow{f} = \{f_1, ..., f_L\}$ : 
$$ P(y=c|\overrightarrow{f}) = \frac{P(y=c)P(\overrightarrow{f}|y=c)}{P(f)}$$

Etant donnée plusieurs classes $c_j$, la classe choisie $\hat{c}$ est celle avec la probabilité maximale 
$$\hat{c} = \arg\max\limits_{c_k} P(y=c_k|\overrightarrow{f})$$
$$\hat{c} = \arg\max\limits_{c_k} \frac{P(y=c_k)P(\overrightarrow{f}|y=c_k)}{P(f)}$$
On supprime l'évidence pour cacher le crime : $P(f)$ ne dépend pas de $c_k$ et elle est postive, donc ça ne va pas affecter la fonction $\max$.
$$\hat{c} = \arg\max\limits_{c_k} P(y=c_k)P(\overrightarrow{f}|y=c_k)$$

Pour calculer $P(\overrightarrow{f}|y=c_k)$, on va utiliser une properiété naïve (d'où vient le nom Naive Bayes) : on suppose l'indépendence conditionnelle entre les caractéristiques $f_j$. 
$$\hat{c} = \arg\max\limits_{c_k} P(y=c_k) \prod\limits_{f_j \in \overrightarrow{f}} P(f_j|y=c_k)$$

Pour éviter la disparition de la probabilité (multiplication et représentation de virgule flottante sur machine), on transforme vers l'espace logarithme.
$$\hat{c} = \arg\max\limits_{c_k} \log P(y=c_k) + \sum\limits_{f_j \in \overrightarrow{f}} \log P(f_j|y=c_k)$$


## I. Réalisation des algorithmes

Pour estimer la vraisemblance, il existe plusieurs modèles (lois):
- **Loi multinomiale :** pour les caracétristiques nominales
- **Loi de Bernoulli :** lorsqu'on est interressé par l'apparence d'une caractéristique ou non (binaire)
- **Loi normale :** pour les caractéristiques numériques

Dans ce TP, nous allons implémenter Naive Bayes pour les caractéristiques nominales (loi multinomiale). 
Dans notre modèle, nous voulons stocker les statistiques et pas les probabilités. 
L'intérêt est de faciliter la mise à jours des statistiques (si par exemple, nous avons un autre dataset et nous voulons enrichir le modèle ; dans e cas, il suffit d'ajouter les statistiques du nouveau dataset)

Ici, nous allons utiliser le dataset "jouer" (utilisé dans la plupart des cours) contenant des caractéristiques nominales.

In [8]:
jouer   = pd.read_csv('data/jouer.csv')

X_jouer = jouer.iloc[:, :-1].values # Premières colonnes 
Y_jouer = jouer.iloc[:,  -1].values # Dernière colonne 

# Afficher le dataset "jouer"
print(jouer)

        temps temperature humidite vent jouer
0   ensoleile      chaude    haute  non   non
1   ensoleile      chaude    haute  oui   non
2     nuageux      chaude    haute  non   oui
3    pluvieux       douce    haute  non   oui
4    pluvieux     fraiche  normale  non   oui
5    pluvieux     fraiche  normale  oui   non
6     nuageux     fraiche  normale  oui   oui
7   ensoleile       douce    haute  non   non
8   ensoleile     fraiche  normale  non   oui
9    pluvieux       douce  normale  non   oui
10  ensoleile       douce  normale  oui   oui
11    nuageux       douce    haute  oui   oui
12    nuageux      chaude  normale  non   oui
13   pluvieux       douce    haute  oui   non


In [9]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


### I.1. Entraînement de la probabilité antérieure

Etant donné le vecteur de sortie $Y$, la probabilité de chaque classe (différentes valeurs de $Y$) est calulée comme :

$$p(c_k) = \frac{|\{y / y \in Y \text{ et } y = c_k\}|}{|Y|}$$


La fonction doit récupérer des statistiques afin de pouvoir calculer la probabilité antérieure de chaque classe. Donc, elle doit retourner  :
- Un vecteur contenant les noms des classes
- Un vecteur contenant les nombres d'occurrences de chaque classe dans le premier vecteur

In [10]:
# TODO: Stastistiques sur la probabilité antérieure
def stat_anterieure(Y: np.ndarray) -> Tuple[np.ndarray, np.ndarray]: 
    cls  = np.unique(Y) # le vecteur des classes
    freq = []
    # compléter à partir d'ici
    cls, freq = np.unique(Y, return_counts=True)
    return cls, np.array(freq)

#=====================================================================
# TEST UNITAIRE
#=====================================================================
# Resultat : 
# (array(['non', 'oui'], dtype=object), array([5, 9]))
#---------------------------------------------------------------------

stat_anterieure(Y_jouer)

(array(['non', 'oui'], dtype=object), array([5, 9]))

### I.2. Entraînement de la probabilité de vraissemblance (loi multinomiale)

Notre modèle doit garder le nombre des différentes valeurs d'une caractéristique $A$ et le nombre de ces valeurs dans chaque classe.
Donc, étant donné un vecteur d'une caractéristique $A= X[:,j]$, un autre des $Y$ et un $C$ contenant la liste des classes, la fonction d'entraînement doit retourner : 
- $V$ : un vecteur contenant les différentes catégories de $A$ (c'est déjà fait)
- Une matrice contenant le nombre d'occurrences de chaque catégorie de $V$ dans chaque classe  : 
   - Les lignes représentent les catégories $v \in V$ de la caréctéristique $A$
   - Les colonnes représentent les classes $c \in C$ de $Y$


In [11]:
# TODO: Statistiques de vraissemblance (une seule caractéristique)
def stat_vraissemblance_1(A: np.ndarray, 
                          Y: np.ndarray, 
                          C: np.ndarray
                         ) -> Tuple[np.ndarray, np.ndarray]: 
    freq = []
    V = np.unique(A) # Catégories de la caractéristique A
    # compléter à partir d'ici

    freq = np.zeros((len(V), len(C)), dtype=np.int32) 
    for i, c in enumerate(C):
        W = A[Y == c] 
        for j, v in enumerate(V):
            freq[j, i] = np.sum(W == v)
    return V, np.array(freq)

#=====================================================================
# TEST UNITAIRE
#=====================================================================
# Resultat : 
# (array(['ensoleile', 'nuageux', 'pluvieux'], dtype=object),
#  array([[3, 2],
#         [0, 4],
#         [2, 3]]))
#---------------------------------------------------------------------
C_t = np.array(['non', 'oui'])
stat_vraissemblance_1(X_jouer[:, 0], Y_jouer, C_t)

(array(['ensoleile', 'nuageux', 'pluvieux'], dtype=object),
 array([[3, 2],
        [0, 4],
        [2, 3]], dtype=int32))

### I.3. Entraînement loi multinomiale

**Rien à programmer ici**

Notre modèle ($\theta_{X, C}$) doit garder des statistiques sur les classes et aussi sur chaque catégorie de chaque caractéristique. Pour ce faire, nous allons représenter $\theta$ comme un vecteur : 
- $\theta[N+1]$ est un vecteur de $N$ éléments représentant des statistiques sur chaque caractéristique $j$, plus un élément (le dernier) pour les statistiques sur les classes.
- Chaque élément est un dictionnaire (HashMap en Java)
- Un élément des caractéristiques contient deux clés : 
    - **val** : pour récupérer la liste des noms des catégories de la caractéristique
    - **freq**: pour récupérer une matrice représentant la fréquence de chaque caractéristique dans chaque classe
- Un élément des classes contient deux clés : 
    - **cls** : pour récupérer la liste des noms des classes
    - **freq**: pour récupérer la liste des fréquences de chaque classe

In [12]:
# La fonction qui entraine Théta sur plusieurs caractéristiques
# Rien à programmer ici
# Notre théta est une liste des dictionnaires;
# chaque dictionnaire contient la liste des catégories et la matrice des fréquences dela caractéristique respective à la colonne de X
# On ajoute les statistiques antérieures des classes à la fin de résultat
def entrainer_multi(X: np.ndarray, 
                    Y: np.ndarray
                   ) -> np.ndarray: 
    
    Theta   = []
    
    stats_c = {}
    stats_c['cls'], stats_c['freq'] =  stat_anterieure(Y)
    
    for j in range(X.shape[1]): 
        stats = {}
        stats['val'], stats['freq'] =  stat_vraissemblance_1(X[:, j], Y, stats_c['cls'])
        Theta.append(stats)
    
    Theta.append(stats_c)
    return Theta


#=====================================================================
# TEST UNITAIRE
#=====================================================================
# Resultat : 
# [{'val': array(['ensoleile', 'nuageux', 'pluvieux'], dtype=object),
#   'freq': array([[3, 2],
#          [0, 4],
#          [2, 3]])},
#  {'val': array(['chaude', 'douce', 'fraiche'], dtype=object),
#   'freq': array([[2, 2],
#          [2, 4],
#          [1, 3]])},
#  {'val': array(['haute', 'normale'], dtype=object),
#   'freq': array([[4, 3],
#          [1, 6]])},
#  {'val': array(['non', 'oui'], dtype=object),
#   'freq': array([[2, 6],
#          [3, 3]])},
#  {'cls': array(['non', 'oui'], dtype=object), 'freq': array([5, 9])}]
#---------------------------------------------------------------------
Theta_jouer = entrainer_multi(X_jouer, Y_jouer)

Theta_jouer

[{'val': array(['ensoleile', 'nuageux', 'pluvieux'], dtype=object),
  'freq': array([[3, 2],
         [0, 4],
         [2, 3]], dtype=int32)},
 {'val': array(['chaude', 'douce', 'fraiche'], dtype=object),
  'freq': array([[2, 2],
         [2, 4],
         [1, 3]], dtype=int32)},
 {'val': array(['haute', 'normale'], dtype=object),
  'freq': array([[4, 3],
         [1, 6]], dtype=int32)},
 {'val': array(['non', 'oui'], dtype=object),
  'freq': array([[2, 6],
         [3, 3]], dtype=int32)},
 {'cls': array(['non', 'oui'], dtype=object), 'freq': array([5, 9])}]

### I.4. Estimation de la probabilité de vraissemblance (loi multinomiale)
L'équation pour estimer la vraisemblance 
$$ P(X_j=v|y=c_k) = \frac{|\{ y \in Y / y = c_k \text{ et } X_j = v\}|}{|\{y = c_k\}|}$$


Dans le cas d'une valeur $v$ qui n'existe pas dans le dataset d'entrainnement ou qui n'existe pas pour une classe donnée mais ui existe dans le dataset de test, nous aurons une probabilité nulle. 
Afin de régler ce problème, nous pouvons appliquer une fonction de lissage qui attribue une petite probabilité aux données non vues dans l'entraînement. 
Le lissage que nous allons utiliser est celui de Lidstone. 
Lorsque $\alpha = 1$, il est appelé lissage de Laplace.
$$ P(X_j=v|y=c_k) = \frac{|\{ y \in Y / y = c_k \text{ et } X_j = v\}| + \alpha}{|\{y = c_k\}| + \alpha * |V|}$$
Où: 
- $\alpha$ est une valeur donnée 
- $V$ est l'ensemble des différentes valeurs de $f_j$ (le vocabulaire; les catégories)

Etant donné : 
- $\theta_j$ les paramètres de la caractéristique $j$ représentées comme dictionnaire
    - **val** : pour récupérer la liste des noms des catégories de la caractéristique (vocabulaire $V$)
    - **freq**: pour récupérer une matrice représentant la fréquence de chaque caractéristique dans chaque classe. C'est une matrice $|V|\times|C|$
- $v$ la valeur de la caractéristique $j$ utilisée pour calculer les probabilités
- $\theta_c$ les paramètres des classes $C$ représentées comme dictionnaire
    - **cls** : pour récupérer la liste des noms des classes
    - **freq**: pour récupérer la liste des fréquences des classes
    
Cette fonction doit retourner : 
- Une liste $P[|C|]$ contenant les probabilités de la catégorie $v$ de $X_j$ sur toutes les classes $C$ 
- Elle doit prendre en considération le cas où la valeur $v$ n'existe pas dans le modèle entraîné

In [13]:
# TODO: Calculer la vraissamblance d'une valeur donnée
def P_vraiss_multi(Theta_j: Dict[str, np.ndarray], 
                   Theta_c: Dict[str, np.ndarray], 
                   v      : str, 
                   alpha  : float = 0.
                  ) -> Tuple[np.ndarray, np.ndarray]: 
    
    #une liste des indices où se trouve la valeur v dans Theta_j["val"]
    ind = np.where(Theta_j['val'] == v)[0] 
    # Récupérer la matrice de fréquences des caractéristiques pour chaque classe
    Matfreq = Theta_j['freq']
    # Récupérer la liste des fréquences des classes
    freqClas = Theta_c['freq']
    # compléter à partir d'ici
    probas = np.zeros(len(Theta_c['cls']))
    # Calculer les probabilités pour chaque classe
    for i, c in enumerate(Theta_c['cls']):
        # Récupérer la fréquence de la valeur v pour la classe c
        if ind.size == 0: # Si la valeur v n'existe pas dans le dataset d'entraînement
            freq_jc_vc = 0
        else:
            freq_jc_vc = Matfreq[ind, i]
        
        # Calculer la probabilité en utilisant le lissage de Lidstone
        probas[i] = (freq_jc_vc + alpha) / (freqClas[i] + alpha * len(Theta_j['val']))
        
    return probas
   
#=====================================================================
# TEST UNITAIRE
#=====================================================================
# Resultat : 
# (array([0.4       , 0.33333333]), array([0.125     , 0.08333333]))
#---------------------------------------------------------------------
# Calcul :
# La probabilité de jouer si temps = pluvieux 
# P(temps = pluvieux | jouer=oui) = (nbr(temps=pluvieux et jouer=oui)+alpha)/(nbr(jour=oui) + alpha * nbr_diff(temps)))
# P(temps = pluvieux | jouer=oui) = (3 + 0)/(9 + 0) ==> 3 est le nombre de différentes valeurs de temps (entrainnement)
# P(temps = pluvieux | jouer=oui) = 4/12 ==> 0.33333333333333333333333333333333333~

# La probabilité de jouer si temps = neigeux 
# P(temps = neigeux | jouer=oui) = (nbr(temps=neigeux et jouer=oui)+alpha)/(nbr(jouer=oui) + alpha * nbr_diff(temps)))
# P(temps = neigeux | jouer=oui) = (0 + 1)/(9 + 3) ==> 3 est le nombre de différentes valeurs de temps (entrainnement)
# P(temps = neigeux | jouer=oui) = 1/13 ==> 0.0833333333333333333333333333333333333~
#---------------------------------------------------------------------

P_vraiss_multi(Theta_jouer[0], Theta_jouer[-1], 'pluvieux'), \
P_vraiss_multi(Theta_jouer[0], Theta_jouer[-1], 'neigeux', alpha=1.)

(array([0.4       , 0.33333333]), array([0.125     , 0.08333333]))

### I.5. Prédiction de la classe (loi multinomiale)
Revenons maintenant à notre équation de prédiction 
$$\hat{c} = \arg\max\limits_{c_k} \log P(y=c_k) + \sum\limits_{f_j \in \overrightarrow{f}} \log P(f_j|y=c_k)$$

- On doit prédire un seule échantillon $x$. 
- La fonction doit retourner un vecteur des log-probabilité des classes
- Si anter=false donc on n'utilise pas la probabilité antérieure

In [14]:
# TODO: Prediction of log probabilities
def predire(x: np.ndarray, Theta: List[Dict[str, np.ndarray]], alpha: float = 1., anter: bool = True) -> float:
    p = np.zeros(2)
    if anter:
        p = np.log(Theta[-1]['freq'] / np.sum(Theta[-1]['freq'][:]))
    for i in range(len(Theta)-1):
        p += np.log(P_vraiss_multi(Theta[i], Theta[-1], x[i], alpha))
    return p

#=====================================================================
# TEST UNITAIRE
#=====================================================================
# Resultat : 
# (array([-5.20912179, -4.10264337]), array([-4.17950237, -3.66081061]))
#---------------------------------------------------------------------
predire(['pluvieux', 'fraiche', 'normale', 'oui'], Theta_jouer), \
predire(['pluvieux', 'fraiche', 'normale', 'oui'], Theta_jouer, anter=False) 

(array([-5.20912179, -4.10264337]), array([-4.17950237, -3.66081061]))

### I.6. Regrouper en une classe (loi multinomiale)

**Rien à programmer ici**


In [ ]:
class NBMultinom(object): 
    
    def __init__(self, alpha=1.): 
        self.alpha = alpha
        
    def entrainer(self, X, Y):
        self.Theta = entrainer_multi(X, Y)
    
    def predire(self, X, anter=True, prob=False): 
        Y_pred = []
        cls = self.Theta[-1]['cls']
        for i in range(len(X)): 
            log_prob = predire(X[i,:], self.Theta, alpha=self.alpha, anter=anter)
            if prob:
                Y_pred.append(np.max(log_prob))
            else:
                Y_pred.append(cls[np.argmax(log_prob)])
        return Y_pred

#=====================================================================
# TEST UNITAIRE
#=====================================================================
# Resultat : 
# ['oui', 'non']
#---------------------------------------------------------------------
notre_modele = NBMultinom()
notre_modele.entrainer(X_jouer, Y_jouer)
X_test = np.array([
    ['neigeux', 'fraiche', 'normale', 'oui'],
    ['neigeux', 'fraiche', 'haute'  , 'oui']
])
notre_modele.predire(X_test)

['oui', 'non']

## II. Application et analyse

**Il n'y a rien à programmer ici.**

Le but de cette section est de mener des expérimentations afin de bien comprendre les concepts vus dans le cours.
Aussi, elle nous assiste à comprendre l'effet des différents paramètres.
En plus, la discussion des différentes expérimentations peut améliorer l'aspect analytique chez l'étudient.

### II.1. Probabilité antérieure 

Nous voulons tester l'effet de la probabilité antérieure.
Pour ce faire, nous avons entraîné deux modèles :
1. Avec probabilité antérieure
1. Sans probabilité antérieure (Il considère une distribution uniforme des classes)

Pour tester si les modèles ont bien s'adapter au dataset d'entraînement, nous allons les tester sur le même dataset et calculer le rapport de classification.


In [15]:
# AVEC Scikit-learn
# ===================
from sklearn.naive_bayes   import CategoricalNB
from sklearn.preprocessing import OrdinalEncoder
from sklearn.metrics       import classification_report

nb_avec     = CategoricalNB(alpha=1.0, fit_prior=True )
nb_sans     = CategoricalNB(alpha=1.0, fit_prior=False)

enc         = OrdinalEncoder()
X_jouer_enc = enc.fit_transform(X_jouer)
nb_avec.fit(X_jouer_enc, Y_jouer)
nb_sans.fit(X_jouer_enc, Y_jouer)

Y_pred_avec = nb_avec.predict(X_jouer_enc)
Y_pred_sans = nb_sans.predict(X_jouer_enc)

# AVEC notre modèle (juste pour voir comment l'utiliser)
# =======================================================
#notre_modele = NBMultinom()
#notre_modele.entrainer(X_jouer, Y_jouer)
#Y_notre_ant = notre_modele.predire(X_jouer)
#Y_notre_sans_ant = notre_modele.predire(X_jouer, anter=False) 

# Le rapport de classification


print( 'Avec probabilité antérieure (a priori)'  )
print(classification_report(Y_jouer, Y_pred_avec))

print( 'Sans probabilité antérieure (a priori)'  )
print(classification_report(Y_jouer, Y_pred_sans))

Avec probabilité antérieure (a priori)
              precision    recall  f1-score   support

         non       1.00      0.80      0.89         5
         oui       0.90      1.00      0.95         9

    accuracy                           0.93        14
   macro avg       0.95      0.90      0.92        14
weighted avg       0.94      0.93      0.93        14

Sans probabilité antérieure (a priori)
              precision    recall  f1-score   support

         non       0.67      0.80      0.73         5
         oui       0.88      0.78      0.82         9

    accuracy                           0.79        14
   macro avg       0.77      0.79      0.78        14
weighted avg       0.80      0.79      0.79        14



**TODO: Analyser les résultats**
    
- Que remarquez-vous ?
- Est-ce que la probabilité antérieure est importante dans ce cas ?
- Comment cette probabilité affecte le résultat ?
- Quand est-ce que nous sommes sûrs que l'utilisation de cette probabilité est inutile ?

**Réponse**

- En comparant les deux rapports de classification, nous pouvons remarquer que le modèle qui utilise la probabilité antérieure (a priori) a une précision plus élevée et un score F1 plus élevé pour les deux classes. En outre, le modèle avec la probabilité antérieure a une précision de 100% pour la classe "non", tandis que le modèle sans la probabilité antérieure a une précision de seulement 67%.
- D'apres la reponse précédente , on peut déduire  que la probabilité antérieure est importante dans ce cas,car elle  est utilisée pour informer le modèle de la probabilité de chaque classe avant de voir les données d'entraînement.
- La probabilité antérieure affecte la façon dont le modèle calcule les probabilités pour chaque classe. Lorsqu'elle est bien choisie, elle peut améliorer les performances du modèle en évitant le sur-apprentissage et en aidant à généraliser les résultats. Cependant, si elle est mal choisie, elle peut affecter négativement les performances du modèle.
- L'utilisation de la probabilité antérieure est souvent considérée comme inutile lorsque les données sont suffisamment représentatives et que les classes sont équilibrées.

### II.2. Lissage

Nous voulons tester l'effet de lissage de Lidstone.
Pour ce faire, nous avons entraîné trois modèles : 
1. alpha = 1 (lissage de Laplace)
1. alpha = 0.5
1. alpha = 0 (sans lissage)

In [18]:
NBC_10 = CategoricalNB(alpha = 1.0 )
NBC_05 = CategoricalNB(alpha = 0.5 )
NBC_00 = CategoricalNB(alpha = 0.0 )

NBC_10.fit( X_jouer_enc,   Y_jouer )
NBC_05.fit( X_jouer_enc,   Y_jouer )
NBC_00.fit( X_jouer_enc,   Y_jouer )

Y_10   = NBC_10.predict(X_jouer_enc)
Y_05   = NBC_05.predict(X_jouer_enc)
Y_00   = NBC_00.predict(X_jouer_enc)


print(          'Alpha = 1.0'             )
print(classification_report(Y_jouer, Y_10))

print(          'Alpha = 0.5'             )
print(classification_report(Y_jouer, Y_05))

print(          'Alpha = 0.0'             )
print(classification_report(Y_jouer, Y_00))


Alpha = 1.0
              precision    recall  f1-score   support

         non       1.00      0.80      0.89         5
         oui       0.90      1.00      0.95         9

    accuracy                           0.93        14
   macro avg       0.95      0.90      0.92        14
weighted avg       0.94      0.93      0.93        14

Alpha = 0.5
              precision    recall  f1-score   support

         non       1.00      0.80      0.89         5
         oui       0.90      1.00      0.95         9

    accuracy                           0.93        14
   macro avg       0.95      0.90      0.92        14
weighted avg       0.94      0.93      0.93        14

Alpha = 0.0
              precision    recall  f1-score   support

         non       1.00      0.80      0.89         5
         oui       0.90      1.00      0.95         9

    accuracy                           0.93        14
   macro avg       0.95      0.90      0.92        14
weighted avg       0.94      0.93     

/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:629: FutureWarning: The default value for `force_alpha` will change to `True` in 1.4. To suppress this warning, manually set the value of `force_alpha`.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:635: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10. Use `force_alpha=True` to keep alpha unchanged.
  warnings.warn(


**TODO: Analyser les résultats**

- Que remarquez-vous ?
- Est-ce que le lissage affecte la performance dans ce cas ? Pourquoi ?
- Pourquoi Scikit-learn n'accepte pas la valeur $\alpha=0$ et affiche une alerte "UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10" ?
- Quelle est l'intérêt du lissage (dans le cas général) ?

**Réponse**

- En analysant les résultats, nous remarquons que les trois modèles ont des performances similaires, avec des scores de précision, de rappel et de f1-score élevés pour les deux classes (non et oui).
- D'apres ce qu'on a remarqué,il est clair que dans ce cas spécifique, le lissage n'a pas affecté les performances du modèle.Cela peut être dû au fait que le jeu de données est relativement petit et simple, avec seulement deux classes et peu de caractéristiques.
- Le lissage de Laplace (ou Lidstone) est une technique courante pour éviter les problèmes de probabilités nulles dans les modèles de classification naïve bayésienne. Cette technique ajoute une petite constante, appelée alpha, à toutes les fréquences de comptage pour éviter que certaines probabilités ne soient égales à zéro.Lorsque alpha est égal à zéro, cela signifie qu'aucun lissage n'est effectué, ce qui peut conduire à des probabilités nulles et à des erreurs numériques dans les calculs. Pour éviter ces erreurs, Scikit-learn fixe la valeur minimale d'alpha à 1e-10.
- Dans le cas général, l'intérêt du lissage est de régulariser les estimations des probabilités conditionnelles pour des classes ou des attributs qui peuvent avoir peu d'exemples dans l'ensemble d'entraînement. Le lissage permet de donner une estimation plus robuste des probabilités et ainsi d'améliorer la généralisation du modèle

### II.3. Comparaison avec d'autres algorithmes

Naive Bayes est un algorithme puissant lorsqu'il s'agit de classer les documents textuels ; nous voulons tester cette information avec la détection de spam. 
Le dataset utilisé est [SMS Spam Collection Dataset](https://www.kaggle.com/uciml/sms-spam-collection-dataset).
Chaque message du dataset doit être représenté sous forme d'un modèle "Sac à mots" (BoW : Bag of Words).
Dans l'entraînement, les différents mots qui s'apparaissent dans les messages (vocabulaire) sont considérés comme des caractéristiques. 
Donc, pour chaque message, la valeur de la caractéristique est la fréquence du mot dans le message. 
Par exemple, si le mot "good" apparait 3 fois dans le message, donc la caractéristique "good" aura la valeur 3 dans ce message.

Notre implémentation n'est pas adéquate pour la nature de ce problème. 
Dans Scikit-learn, [sklearn.naive_bayes.CategoricalNB](https://scikit-learn.org/stable/modules/generated/sklearn.naive_bayes.CategoricalNB.html) est similaire à notre implémentation. 
L'algorithme adéquat pour ce type de problème est [sklearn.naive_bayes.MultinomialNB](https://scikit-learn.org/stable/modules/generated/sklearn.naive_bayes.MultinomialNB.html).
Les algorithmes comparés :
1. Naive Bayes (Loi Multinomiale)
1. Naive Bayes (Loi Gaussienne)
1. Regression logistique 
1. Arbres de decision

In [19]:
# lire le dataset
messages = pd.read_csv('data/spam.csv', encoding='latin-1')
# renomer les caractéristiques : texte et classe
messages = messages.rename(columns={'v1': 'classe', 'v2': 'texte'})
# garder seulement ces deux caractéristiques
messages = messages.filter(['texte', 'classe'])

messages.head()

texte classe
0  Go until jurong point, crazy.. Available only ...    ham
1                      Ok lar... Joking wif u oni...    ham
2  Free entry in 2 a wkly comp to win FA Cup fina...   spam
3  U dun say so early hor... U c already then say...    ham
4  Nah I don't think he goes to usf, he lives aro...    ham

In [20]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.model_selection         import train_test_split
from sklearn.naive_bayes             import MultinomialNB, GaussianNB
from sklearn.linear_model            import LogisticRegression
from sklearn.tree                    import DecisionTreeClassifier
from sklearn.metrics                 import precision_score, recall_score
import timeit


modeles = [
    MultinomialNB(),
    GaussianNB(),
    LogisticRegression(solver='lbfgs') ,
    #solver=sag est plus lent; donc j'ai choisi le plus rapide
    DecisionTreeClassifier()
]

temps_train = []
temps_test  = []
rappel      = []
precision   = []

msg_train, msg_test, Y_train, Y_test = train_test_split(messages['texte'] ,
                                                        messages['classe'],
                                                        test_size    = 0.2, 
                                                        random_state = 0  )

count_vectorizer = CountVectorizer()
X_train          = count_vectorizer.fit_transform(msg_train).toarray()
X_test           = count_vectorizer.transform    (msg_test ).toarray()


for modele in modeles:
    # ==================================
    # ENTRAINEMENT 
    # ==================================
    temps_debut = timeit.default_timer()
    modele.fit(X_train, Y_train)
    temps_train.append(timeit.default_timer() - temps_debut)
    
    # ==================================
    # TEST 
    # ==================================
    temps_debut = timeit.default_timer()
    Y_pred      = modele.predict(X_test)
    temps_test.append(timeit.default_timer() - temps_debut)
    
    # ==================================
    # PERFORMANCE 
    # ==================================
    # Ici, nous considérons une classification binaire avec une seule classe "spam" 
    # le classifieur ne sera pas jugé par sa capacité de détecter les non spams
    precision.append(precision_score(Y_test, Y_pred, pos_label='spam'))
    rappel   .append(recall_score   (Y_test, Y_pred, pos_label='spam'))

    
print('Fin') 

Fin


#### II.3.1. Temps d'entraînement et de test

Combien de temps chaque algorithme prend pour entrainer le même dataset d'entrainement et combien de temps pour tester le même dataset de test.

In [23]:
algo_noms = ['Naive Bayes Multinomial (NBM)', 
             'Naive Bayes Gaussien (NBG)'   , 
             'Regression logistique (RL)'   , 
             'Arbre de decision (AD)']

pd.DataFrame({
    'Algorithme'            : algo_noms  ,
    'Temps d\'entrainement' : temps_train,
    'Temps de test'         : temps_test
})

Algorithme  Temps d'entrainement  Temps de test
0  Naive Bayes Multinomial (NBM)              0.976525       0.061037
1     Naive Bayes Gaussien (NBG)              0.817354       0.150747
2     Regression logistique (RL)              4.115931       0.073569
3         Arbre de decision (AD)             22.159051       0.027639

**TODO: Analyser les résultats**

- Que remarquez-vous concernant le temps d'entrainement ? (ordonner les algorithmes)
- Pourquoi nous avons eu ces résultats en se basant sur les algorithmes ? (discuter chaque algorithme vis-a-vis le temps d'entrainement)
- Que remarquez-vous concernant le temps de test ? (ordonner les algorithmes)
- Pourquoi nous avons eu ces résultats en se basant sur les algorithmes ? (discuter chaque algorithme vis-a-vis le temps de test)

**Réponse**

- Concernant le temps d'entraînement, nous remarquons que l'algorithme le plus rapide estNaive Bayes Gaussien (NBG)	, suiviNaive Bayes Multinomial (NBM)		, la Regression logistique (RL), puis enfin Arbre de decision (AD)	 qui est le plus lent.
- Ces résultats sont attendus car chaque algorithme a une complexité différente en termes de temps d'entraînement. Les algorithmes de type Bayesien, tels que le Naive Bayes Gaussien et le Naive Bayes Multinomial, sont généralement plus rapides à entraîner car ils ont une complexité linéaire en fonction de la taille des données d'entraînement. La Régression logistique a une complexité légèrement supérieure car elle doit optimiser des coefficients pour ajuster le modèle aux données. Enfin, l'Arbre de décision a une complexité plus élevée car il doit chercher la meilleure partition pour chaque nœud de l'arbre, ce qui peut être très coûteux pour les grands ensembles de données.
- Concernant le temps de test, nous remarquons que l'algorithme le plus rapide est l'arbre de décision (AD) avec un temps de test de 0.027639 secondes, suivi par Naive Bayes Multinomial (NBM) avec 0.061037 secondes, la Regression logistique (RL) avec 0.073569 secondes, et enfin Naive Bayes Gaussien (NBG) avec 0.150747 secondes.
- Ces résultats s'expliquent en partie par les caractéristiques de chaque algorithme. L'Arbre de décision (AD) est un algorithme relativement simple qui consiste à diviser de manière récursive les données en fonction de seuils sur les différentes variables, ce qui rend son temps de test très rapide. De même, la Régression logistique (RL) est également un algorithme assez simple qui peut être résolu de manière analytique, ce qui en fait un algorithme rapide en temps de test.En revanche, les deux algorithmes de Naive Bayes nécessitent de calculer des probabilités, ce qui peut être plus coûteux en temps de calcul. Le Naive Bayes Multinomial (NBM) nécessite de calculer les probabilités d'apparition de chaque mot dans chaque catégorie, tandis que le Naive Bayes Gaussien (NBG) nécessite de calculer les moyennes et les variances de chaque variable pour chaque catégorie. Ces calculs peuvent être plus coûteux en temps de test que les algorithmes simples comme l'AD ou la RL.


#### II.3.2. Qualité de prédiction

Comment chaque algorithme performe sur le dataset de test dans le cas de détection de spams (spam: est la classe positive).

In [22]:
pd.DataFrame({
    'Algorithme' : algo_noms,
    'Rappel'     : rappel   ,
    'Precision'  : precision
})

Algorithme    Rappel  Precision
0  Naive Bayes Multinomial (NBM)  0.927711   0.987179
1     Naive Bayes Gaussien (NBG)  0.891566   0.616667
2     Regression logistique (RL)  0.855422   0.986111
3         Arbre de decision (AD)  0.825301   0.907285

**TODO: Analyser les résultats**

On remarque que Naive Bayes surpasse la régression logistique pour la détection de spams. 
- Est-ce que ceci preuve que Naive Bayes est meilleur que les autres algorithmes sur n'importe quel problème ?
- Est-ce que ceci preuve que Naive Bayes peut donner de meilleurs résultats que les autres algorithmes sur des problèmes similaires ?
- Pourquoi le modèle gaussien est moins performant que le multinomial en se basant sur la nature des deux algorithmes ?
- Pourquoi le modèle gaussien est moins performant que le multinomial en se basant sur la nature du probleme/donnees ?

**Réponse**

- 
Non, cela ne prouve pas que Naive Bayes est meilleur que les autres algorithmes sur n'importe quel problème.
- Oui, Naive Bayes peut donner de meilleurs résultats que les autres algorithmes sur des problèmes similaires
- Le modèle gaussien est moins performant que le modèle multinomial dans la détection de spams car il suppose que les variables d'entrée suivent une distribution normale. Cependant, dans le cas de la détection de spams, les variables d'entrée sont des fréquences d'apparition de mots, ce qui ne suit pas une distribution normale. Le modèle gaussien ne peut donc pas capturer toutes les nuances des données et ne convient pas bien pour ce type de problème.En revanche, le modèle multinomial prend en compte la fréquence d'apparition des mots, ce qui est plus adapté pour ce type de données.
-  Les modèles de type multinomial sont plus adaptés pour les données discrètes, comme les données de texte dans ce cas, où chaque mot peut être considéré comme une variable discrète représentant la fréquence d'apparition d'un mot dans un message. En revanche, le modèle gaussien suppose que les variables sont continues et suivent une distribution gaussienne, ce qui est moins adapté aux données de texte.

In [16]:
print("  _____    __                                              _               ")
print(" |_   _|  / _|                                            | |              ")
print("   | |   | |_     _   _    ___    _   _      __ _    ___  | |_             ")
print("   | |   |  _|   | | | |  / _ \  | | | |    / _` |  / _ \ | __|            ")
print("  _| |_  | |     | |_| | | (_) | | |_| |   | (_| | |  __/ | |_             ")
print(" |_____| |_|      \__, |  \___/   \__,_|    \__, |  \___|  \__|            ")
print("                   __/ |                     __/ |                         ")
print("                  |___/                     |___/                          ")
print("  _     _       _            __                                            ")
print(" | |   | |     (_)          / _|                 _                         ")
print(" | |_  | |__    _   ___    | |_    __ _   _ __  (_)                        ")
print(" | __| | '_ \  | | / __|   |  _|  / _` | | '__|                            ")
print(" | |_  | | | | | | \__ \   | |   | (_| | | |     _                         ")
print("  \__| |_| |_| |_| |___/   |_|    \__,_| |_|    ( )                        ")
print("                                                |/                         ")
print("                                                                           ")
print("                                                                           ")
print("                                                                           ")
print("  _   _    ___    _   _      __ _   _ __    ___                            ")
print(" | | | |  / _ \  | | | |    / _` | | '__|  / _ \                           ")
print(" | |_| | | (_) | | |_| |   | (_| | | |    |  __/                           ")
print("  \__, |  \___/   \__,_|    \__,_| |_|     \___|                           ")
print("   __/ |                                                                   ")
print("  |___/                                                                    ")
print("                    _                                                __    ")
print("                   | |                                            _  \ \   ")
print("  _ __     ___     | |__    _   _   _ __ ___     __ _   _ __     (_)  | |  ")
print(" | '_ \   / _ \    | '_ \  | | | | | '_ ` _ \   / _` | | '_ \         | |  ")
print(" | | | | | (_) |   | | | | | |_| | | | | | | | | (_| | | | | |    _   | |  ")
print(" |_| |_|  \___/    |_| |_|  \__,_| |_| |_| |_|  \__,_| |_| |_|   (_)  | |  ")
print("                                                                     /_/   ")
print("                                                                           ")

  _____    __                                              _               
 |_   _|  / _|                                            | |              
   | |   | |_     _   _    ___    _   _      __ _    ___  | |_             
   | |   |  _|   | | | |  / _ \  | | | |    / _` |  / _ \ | __|            
  _| |_  | |     | |_| | | (_) | | |_| |   | (_| | |  __/ | |_             
 |_____| |_|      \__, |  \___/   \__,_|    \__, |  \___|  \__|            
                   __/ |                     __/ |                         
                  |___/                     |___/                          
  _     _       _            __                                            
 | |   | |     (_)          / _|                 _                         
 | |_  | |__    _   ___    | |_    __ _   _ __  (_)                        
 | __| | '_ \  | | / __|   |  _|  / _` | | '__|                            
 | |_  | | | | | | \__ \   | |   | (_| | | |     _                         
  \__| |_| |